### PR TITLE
Fixes for PyTorch/XLA functionalization integration

### PR DESCRIPTION
**Pull request by wonjoo-wj**
Picking up https://github.com/pytorch/pytorch/pull/88506

### CODE REVIEW

**Comment by alanwaketan** [comment]
@wonjoolee95 You probably want to hide your debug hints while running the CI here. It makes the log too large and impossible to parse over the browser. Also, a rebase will be appreciated.

**Comment by alanwaketan** [comment]
I guess it's better to log those information with proper log level control such that you can still use it locally for debugging but it won't increase the test log size dramatically.

**Comment by wonjoolee95** [comment]
Hmm, seems like some functorch tests are still failing even after applying the diff generated by `EXPECTTEST_ACCEPT=1`. I'll try it one more time. If they're still failing, I'll fix the rest manually.

**Comment by alanwaketan** [comment]
It looks like those machines are with gpu devices. Do you have a gpu env? I guess you need to run those tests on a gpu env. Otherwise, the tests will be skipped.

**Comment by wonjoolee95** [comment]
Please correct me if I'm wrong, seems like only the first two failing tests are gpu devices? I'll wait for the rest of the CI to complete and then first all the non-gpu tests first.

**Comment by wonjoolee95** [comment]
With the latest commit, all the TestAOTAutograd tests should succeed:
```
(base) jenkins@26d7adccbc26:/workspace/pytorch$ python test/functorch/test_aotdispatch.py TestAOTAutograd
/opt/conda/lib/python3.7/site-packages/torchvision/io/image.py:13: UserWarning: Failed to load image Python extension: libc10_cuda.so: cannot open shared object file: No such file or directory
  warn(f"Failed to load image Python extension: {e}")
ss2022-12-15 00:26:27.721580: W 2153590 tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcuda.so.1'; dlerror: libcuda.so.1: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /usr/local/cuda/lib64:/usr/local/nvidia/lib:/usr/local/nvidia/lib64
2022-12-15 00:26:27.721700: W 2153590 tensorflow/compiler/xla/stream_executor/cuda/cuda_driver.cc:265] failed call to cuInit: UNKNOWN ERROR (303)
./opt/conda/lib/python3.7/site-packages/torch/_functorch/aot_autograd.py:919: UserWarning: Your compiler for AOTAutograd is returning a a function that doesn't take boxed arguments. Please wrap it with functorch.compile.make_boxed_func or handle the boxed arguments yourself. See https://github.com/pytorch/pytorch/pull/83137#issuecomment-1211320670 for rationale.
  "Your compiler for AOTAutograd is returning a a function that doesn't take boxed arguments. "
...test/functorch/test_aotdispatch.py:241: UserWarning: The .grad attribute of a Tensor that is not a leaf Tensor is being accessed. Its .grad attribute won't be populated during autograd.backward(). If you indeed want the .grad field to be populated for a non-leaf Tensor, use .retain_grad() on the non-leaf Tensor. If you access the non-leaf Tensor by mistake, make sure you access the leaf Tensor instead. See github.com/pytorch/pytorch/pull/30531 for more informations. (Triggered internally at /workspace/pytorch/build/aten/src/ATen/core/TensorBody.h:485.)
  grads = [inp.grad for inp in pytree.tree_flatten(inps)[0]]
...................................s...../opt/conda/lib/python3.7/site-packages/torch/_functorch/aot_autograd.py:919: UserWarning: Your compiler for AOTAutograd is returning a a function that doesn't take boxed arguments. Please wrap it with functorch.compile.make_boxed_func or handle the boxed arguments yourself. See https://github.com/pytorch/pytorch/pull/83137#issuecomment-1211320670 for rationale.
  "Your compiler for AOTAutograd is returning a a function that doesn't take boxed arguments. "
..../opt/conda/lib/python3.7/site-packages/torch/_functorch/aot_autograd.py:919: UserWarning: Your compiler for AOTAutograd is returning a a function that doesn't take boxed arguments. Please wrap it with functorch.compile.make_boxed_func or handle the boxed arguments yourself. See https://github.com/pytorch/pytorch/pull/83137#issuecomment-1211320670 for rationale.
  "Your compiler for AOTAutograd is returning a a function that doesn't take boxed arguments. "
...
----------------------------------------------------------------------
Ran 54 tests in 3.871s

OK (skipped=3)
(base) jenkins@26d7adccbc26:/workspace/pytorch$ 
```
I'm still unsure about the previous GPU failure that failed with:
```
/var/lib/jenkins/multipy/multipy/runtime/../../multipy/runtime/interpreter/builtin_registry.h:31:10: fatal error: gtest/gtest_prod.h: No such file or directory
```
But I'll let the CI to run one more time before spending more time on the GPU test.

**Comment by alanwaketan** [comment]
That failure is very likely unrelated.

**Comment by alanwaketan** [comment]
According to [hud](https://hud.pytorch.org/), the deploy failure shouldn't be related.

File: torchgen/gen_functionalization_type.py [hunk 1]
```diff
@@ -347,7 +347,6 @@ def emit_view_functionalization_body(
         }}
       );
       auto compute_reference_meta =
-        {view_tensor_name}.key_set().has_backend(c10::BackendComponent::XLABit) ||
```
Comment (alanwaketan) [inline torchgen/gen_functionalization_type.py:350]: @bdhirsh Is this guard safe to remove? With it, I crashed on xla symbolic expand:
```
root@t1v-n-307ffe96-w-0:/workspaces/work/pytorch/xla# PJRT_DEVICE=CPU python test/test_dynamic_shapes.py -v TestDynamicShapes.test_simple_expand
test_simple_expand (__main__.TestDynamicShapes) ... ERROR

======================================================================
ERROR: test_simple_expand (__main__.TestDynamicShapes)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test/test_dynamic_shapes.py", line 18, in test_simple_expand
    t5.expand(t2.size(0))
RuntimeError: /workspaces/work/pytorch/build/aten/src/ATen/RegisterCompositeExplicitAutograd.cpp:2109: SymIntArrayRef expected to contain only concrete integers

----------------------------------------------------------------------
Ran 1 test in 0.022s

FAILED (errors=1)
```

Comment (alanwaketan) [reply torchgen/gen_functionalization_type.py:350]: Okay, it regresses. It's not safe.

Comment (bdhirsh) [reply torchgen/gen_functionalization_type.py]: @alanwaketan so, the idea motivating this bit if code is the following:
* pytorch/XLA doesn't care about strides, so when comparing a pytorch program when run on CUDA vs XLA, the user will witness different strides on the tensors throughout their program
* functionalization gives XLA the ability to fix that problem; XLA can choose to not care about the value of strides in all of its kernels, but functionalization can run the meta function for every ATen, to properly set the strides

One question here is - do you think that's a benefit worth trying to capture for pytorch/XLA (stride correctness, for the user's perspective, for XLA tensors)? I'd be interested in @JackCaoG 's opinion.

Our options are either to:
(1) kill that code, and not bother trying to get correct strides
(2) make it more robust so it works on this test

In this test, it looks like you're using dynamic shapes, and the meta function we're calling doesn't play well with dynamic shapes. The way that the dynamic shapes workstream in core has been handling this is that we have python implementations / decompositons of a bunch of our ops, that we want to run when dynamic shapes are enabled. And it looks like... for some reason we aren't calling that python impl, and are instead calling the C++ one?

There's probably a better way to arrange for this to work with XLA, but one option option is enable the python dispatcher in your test, which should override a bunch of C++ meta kernels with their python equivalents:
```
from torch._dispatch.python import enable_python_dispatcher
with enable_python_dispatcher():
    test()
```

**Comment by wonjoolee95** [comment]
Rebased this and the XLA POC PR with master. The `aodispatch` tests here will fail, as the master branch had some changes as well. I'll follow-up later to fix those.

File: torchgen/gen_functionalization_type.py (hunk 1 above)
Comment (alanwaketan) [reply torchgen/gen_functionalization_type.py]: Here is the follow up on the xla side: https://github.com/pytorch/xla/issues/4448.

Comment (alanwaketan) [reply torchgen/gen_functionalization_type.py]: Looks like by enabling python dispatcher and implementing missing sym size ops can workaround this. But then it brings a bigger question whether we should enable python dispatcher for dynamic shapes or not.

**Comment by wonjoolee95** [comment]
Bunch of `dynamo` related tests failing now with messages:
```
 Error: trace_fork_wait_inline (jit.test_async.TestAsync) ... [2023-01-25 03:37:54,841] torch._dynamo.convert_frame: [ERROR] WON'T CONVERT <graph break in test_trace_fork_wait_inline> /var/lib/jenkins/workspace/test/jit/test_async.py line 416 
```
The last commit does touch dynamo related code but the CI used to be green even with that commit. While I try to reproduce this locally, let me also try to rebase from master and re-run this CI.

**Comment by alanwaketan** [comment]
> Bunch of `dynamo` related tests failing now with messages:
> 
> ```
>  Error: trace_fork_wait_inline (jit.test_async.TestAsync) ... [2023-01-25 03:37:54,841] torch._dynamo.convert_frame: [ERROR] WON'T CONVERT <graph break in test_trace_fork_wait_inline> /var/lib/jenkins/workspace/test/jit/test_async.py line 416 
> ```
> 
> The last commit does touch dynamo related code but the CI used to be green even with that commit. While I try to reproduce this locally, let me also try to rebase from master and re-run this CI.

We can always use hud.pytorch.org to determine if the test is broken in tip of the tree.

**Comment by wonjoolee95** [comment]
Thanks for the info, Jiewen. Looks like master also was seeing the same issue for a while:
Screenshot 2023-01-25 at 1 49 25 PM
But seems like a recent commit fixed it, so this PR's CI should be good now. I'll let the CI verify.

File: aten/src/ATen/native/native_functions.yaml [hunk 2]
```diff
@@ -14580,3 +14580,6 @@
   dispatch:
     CUDA: _fused_adamw_kernel_cuda_
   autogen: _fused_adamw, _fused_adamw.out
+
+- func: _propagate_xla_data(Tensor input, Tensor output) -> ()
```
Comment (alanwaketan) [inline aten/src/ATen/native/native_functions.yaml:14584]: @bdhirsh I have added the op you suggested in https://github.com/pytorch/xla/pull/4505#issuecomment-1405176038.  Please review. You can just check the commit called: [Functionalization] Adds _propagate_xla_data.

**Comment by wonjoolee95** [comment]
Moving to a new PR with a new branch -- https://github.com/pytorch/pytorch/pull/94537. Marking this one closed.

**Pull request closed without merging**